libs/screen/image.cpp

```diff
@@ -1159,8 +1159,191 @@ void _fillCircle(Image_ img, int cxy, int r, int c) {
+        }
+        if (line->D > 0) {
+            line->y += line->yi;
+            line->D -= line->dx;
+        }
+        line->D += line->dy;
+        ++line->x;
+    }
+}
+
+void nextYRange_HighUp(int x, LineGenState *line, ValueRange *yRange) {
+    while (line->x == x && line->y >= line->y1 && line->x < line->W) {
+        if (0 <= line->x) {
+            if (line->y < yRange->min) yRange->min = line->y;
+            if (line->y > yRange->max) yRange->max = line->y;
+        }
+        if (line->D > 0) {
+            line->x += line->xi;
+            line->D += line->dy;
+        }
+        line->D += line->dx;
+        --line->y;
+    }
+}
+
+void nextYRange_HighDown(int x, LineGenState *line, ValueRange *yRange) {
+    while (line->x == x && line->y <= line->y1 && line->x < line->W) {
+        if (0 <= line->x) {
+            if (line->y < yRange->min) yRange->min = line->y;
+            if (line->y > yRange->max) yRange->max = line->y;
+        }
+        if (line->D > 0) {
+            line->x += line->xi;
+            line->D -= line->dy;
+        }
+        line->D += line->dx;
+        ++line->y;
+    }
+}
+
+LineGenState initYRangeGenerator(int16_t X0, int16_t Y0, int16_t X1, int16_t Y1) {
+    LineGenState line;
+
+    line.x0 = X0, line.y0 = Y0, line.x1 = X1, line.y1 = Y1;
+
+    line.dx = line.x1 - line.x0;
+    line.dy = line.y1 - line.y0;
+    line.y = line.y0;
+    line.x = line.x0;
+
+    if ((line.dy < 0 ? -line.dy : line.dy) < line.dx) {
+        line.yi = 1;
+        if (line.dy < 0) {
+            line.yi = -1;
+            line.dy = -line.dy;
+        }
+        line.D = 2 * line.dy - line.dx;
+        line.dx <<= 1;
+        line.dy <<= 1;
+
+        line.nextFuncIndex=0;
+        return line;
+    } else {
+        line.xi = 1;
+        // if (dx < 0) {//should not hit
+        //     PANIC();
+        // }
+        if (line.dy < 0) {
+            line.D = 2 * line.dx + line.dy;
+            line.dx <<= 1;
+            line.dy <<= 1;
+
+            line.nextFuncIndex = 1;
+            return line;
+        } else {
+            line.D = 2 * line.dx - line.dy;
+            line.dx <<= 1;
+            line.dy <<= 1;
+
+            line.nextFuncIndex = 2;
+            return line;
+        }
+    }
+}
+
 void fillTriangle(Image_ img, int x0, int y0, int x1, int y1, int x2, int y2, int c) {
-    // Fill triangle here!
+    if (x1 < x0) {
+        swap(x0, x1);
+        swap(y0, y1);
+    }
+    if (x2 < x1) {
+        swap(x1, x2);
+        swap(y1, y2);
+    }
+    if (x1 < x0) {
+        swap(x0, x1);
+        swap(y0, y1);
+    }
+
+    LineGenState lines[]={
+        initYRangeGenerator(x0, y0, x2, y2),
+        initYRangeGenerator(x0, y0, x1, y1),
+        initYRangeGenerator(x1, y1, x2, y2)};
+    
+    lines[0].W = lines[1].W = lines[2].W = width(img);
+    lines[0].H = lines[1].H= lines[2].H = height(img);
+
+    typedef void (*FP_NEXT)(int x, LineGenState *line, ValueRange *yRange);
```

Comment:
We have 3 diff sub-function to generate Ys of edges, each particular edge map to one of them. With this kind of method pointers storage the entrence of the methods, to avoid judge which one to call at every X.

Reply:
```suggestion
// We have 3 diff sub-function to generate Ys of edges, each particular edge map to one of them. 
// With this kind of method pointers storage the entrence of the methods, to avoid judge which one to call at every X.
    typedef void (*FP_NEXT)(int x, LineGenState *line, ValueRange *yRange);
```